### ISSUE TITLE
Add pixel type support to GDCMImageIO

### ISSUE THREAD

**Issue by zivy** [InsightSoftwareConsortium/ITK#2078]
Not sure if this is a bug or feature request. I am trying to read a segmentation file, [liver.dcm](https://github.com/QIICR/dcmqi/blob/master/data/segmentations/liver.dcm), available from the dcmqi repo. This throws an exception with error message of "Unhandled PixelFormat". GDCM itself is able to successfully read this file.

@fedorov, @pieper, 
Any input with respect to reading a DICOM stored segmentation as an ITK label image appreciated. I am aware that it is possible that a voxel has multiple labels or fractional values and that the segmentation file may not have the same spatial sampling or extent as the referenced image. 

I am thinking of the simplified situation where each voxel has a single label and the spacing/size are encoded in the DICOM segmentation file. Should ITK be able to directly read such a multi-label segmentation, as if it were reading an image?

**Comment by fedorov** [comment]
@zivy you mentioned dcmqi, so I assume you are familiar with how it works - it generates a separate ITK image for each label contained in the DICOM SEG, where the geometry of each of the images is fully defined by the DICOM SEG. Would it be possible for the ITK reader to return an array of images, in a similar way?

> I am thinking of the simplified situation where each voxel has a single label and the spacing/size are encoded in the DICOM segmentation file. Should ITK be able to directly read such a multi-label segmentation, as if it were reading an image?

Yes, I think this would be practical, and better than throwing the error, and it is indeed the case that most segmentations do not have overlapping segments. But it will remain the case that overlapping segments situation would need to be detected. 

IMHO no matter what is the initial implementation, it would be good to plan it in such a way that does not preclude support for overlapping segments in the future.

**Comment by pieper** [comment]
It would be great to have ITK support dicom SEG natively.  It shouldn't be too difficult, especially if you use dcmtk and then use dcmqi as a dependency.

**Comment by zivy** [comment]
@fedorov,

Thanks for the input. What I want to do is read a segmentation the same way as reading an image, returning multiple images would break the ImageIO framework. To make it compatible with image reading I would propose that it return a multi-channel (vector pixel) image, where each channel corresponds to a label.

**Comment by issakomi** [comment]
SINGLEBIT pixel type was recently added to GDCMImageIO, please try master,
s. https://github.com/InsightSoftwareConsortium/ITK/pull/2010

**Comment by issakomi** [comment]
Please note that GDCMImageIO doesn't check geometry of multi-frame images, it reads shared group and 1st frame of per-frame group, so it works if frames have same orientation, are equidistant and correctly sorted. But it is applicable to all multi-frame images, not to above PR. It were not so difficult to add such checks and to give a warnings, not sure such PR were welcome.

**Comment by fedorov** [comment]
> What I want to do is read a segmentation the same way as reading an image, returning multiple images would break the ImageIO framework. 

And I assume it would not be feasible to have a multi-step read process, where you first get the number of segments, and then request read of a single segment at a time?

> I would propose that it return a multi-channel (vector pixel) image, where each channel corresponds to a label

This will probably work, but can be very inefficient in some cases, since your vector image would need to span the bounding box of all segments included.

**Comment by zivy** [comment]
@fedorov,
You're right with respect to the inefficiency of a vector pixel type, wasn't thinking it through. It should probably return a [LabelMap image](https://itk.org/Doxygen/html/classitk_1_1LabelMap.html) which is designed for efficient storage of labels.

**Comment by dzenanz** [comment]
@zivy `itk::ImageFileReader` returns the kind of image with which it was instantiated (the template parameter). And user controls that.

**Comment by blowekamp** [comment]
> `itk::ImageFileReader` returns the kind of image with which it was instantiated (the template parameter). And user controls that.

The ImageFileReader does not support LabelMaps as a type, and there is no support for RLE encoded buffers in the ImageIO classes as far as I know. There is the [RLEImage](https://blog.kitware.com/memory-compression-in-itk-snap/) remote module, that is relevant here, but I don't have experience with.

There may be a need for an new/alternative/modified ImageFileReader class which presents the a list of Images in a file. Perhaps after `UpdateOutputInformation` is called. And allows the selection of one of them ( perhaps defaulting to the first ).

Relevant here is also file formats which store multi-resolution/pyramid images. There has not been a good way to select which resolution of image is loaded with the ITK interface. Some formats (like TIFF) readily support multiple image in the file, sometimes for a thumbnails too.

**Comment by dzenanz** [comment]
`RLEImage` works only if it is being accessed via iterators and GetPixel/SetPixel. There is no GetBufferPointer which readers are probably using.

**Comment by fedorov** [comment]
> there is no support for RLE encoded buffers

Why is this important?

**Comment by issakomi** [comment]
BTW, there is support for RLE compression in GDCMImageIO (read) and one more note (without stepping in into discussion about special type for _Labels_): _single bit_ pixel type in _Segmentation Image Storage_ have to be of type _BINARY_, it is constraint in [C.8.20.2](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.8.20.2.html). _FRACTIONAL_ (_PROBABILITY_ or _OCCUPANCY_) have to have 8 bits allocated (no idea why only 8, BTW) and will be opened in GDCM IO as usual _char_ image currently. That liver.dcm above is _BINARY_ / _single bit_.

![Screenshot at 2020-10-29 03-26-00](https://user-images.githubusercontent.com/1494580/97518227-1f405d00-1997-11eb-91b1-a84b59341f26.png)
